runtime/compiler/env/DependencyTable.hpp

```diff
@@ -0,0 +1,89 @@
+/*******************************************************************************
+ * Copyright IBM Corp. and others 2024
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] https://openjdk.org/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+ *******************************************************************************/
+
+#ifndef DEPENDENCYTABLE_INCL
+#define DEPENDENCYTABLE_INCL
+
+#include "env/J9PersistentInfo.hpp"
+#include "env/PersistentCollections.hpp"
+#include "env/VMJ9.h"
+
+class TR_J9SharedCache;
+
+struct OffsetEntry
+   {
+   PersistentUnorderedSet<J9Class *> _loadedClasses;
+   };
+
+
+/**
+ * \brief Tracking AOT load dependencies
+ *
+ * The dependency table maintains a map from ROM class offsets to loaded
+ * J9Classes with that offset and a valid stored class chain. (There is at most
+ * one chain with a particular first ROM class offset, so any two such classes
+ * must have identical chains). The ROM classes of these classes must have been
+ * shared when they were loaded in order to be tracked.
```

Comment:
It isn't necessary for AOT dependency tracking to track classes whose ROM classes weren't in the SCC when they were loaded - if they weren't, then the only way compiled code in the SCC could reference them is if this JVM was responsible for compiling that code (and remembering the class during the JVM run). If that's the case, we don't need to track the dependencies of that code anyway.

I could have the table include all classes by updating the table in `TR_J9SharedCache::rememberClass()`, but there was one potential race that I was concerned about, where we remember a class that's loaded but not initialized at the same time as the class initialization hook is triggered for that class. We might end up not notifying the table of the initialization properly in that case. I mention this because if the `_offsetMap` here is used to look up candidate classes during relocation in case looking them up in a class loader fails, then we might want freshly-remembered classes in the table to improve that fallback. That's an optimization, though.

Reply:
I meant to say class chains, not ROM classes above. I could also `rememberClass()` on class load instead of using `classMatchesCachedVersion()`, or track every shared class. I believe either would also work.